File: src/ObjectManager/Object.Tes/FilePacks/Records/045-LSCR.Load Screen.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class LSCRRecord : Record
    {
        public struct LNAMField
        {
            public FormId<Record> Direct;
            public FormId<WRLDRecord> IndirectWorld;
            public short IndirectGridX;
            public short IndirectGridY;

            public LNAMField(UnityBinaryReader r, int dataSize)
            {
                Direct = new FormId<Record>(r.ReadLEUInt32());
                //if (dataSize == 0)
                IndirectWorld = new FormId<WRLDRecord>(r.ReadLEUInt32());
                //if (dataSize == 0)
                IndirectGridX = r.ReadLEInt16();
                IndirectGridY = r.ReadLEInt16();
            }
        }

        public override string ToString() => $"LSCR: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public FILEField ICON; // Icon
        public STRVField DESC; // Description
        public List<LNAMField> LNAMs; // LoadForm

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "ICON": ICON = new FILEField(r, dataSize); return true;
                case "DESC": DESC = new STRVField(r, dataSize); return true;
                case "LNAM": if (LNAMs == null) LNAMs = new List<LNAMField>(); LNAMs.Add(new LNAMField(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-LSCR.Load Screen.swift
//
//  LSCRRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class LSCRRecord: Record {
    public struct LNAMField {
        public let direct: FormId<Record>
        public let indirectWorld: FormId<WRLDRecord>
        public let indirectGridX: Int16
        public let indirectGridY: Int16

        init(_ r: BinaryReader, _ dataSize: Int) {
            direct = FormId<Record>(r.readLEUInt32())
            indirectWorld = FormId<WRLDRecord>(r.readLEUInt32())
            indirectGridX = r.readLEInt16()
            indirectGridY = r.readLEInt16()
        }
    }

    public override var description: String { return "LSCR: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var ICON: FILEField! // Icon
    public var DESC: STRVField! // Description
    public var LNAMs: [LNAMField]! // LoadForm
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "ICON": ICON = r.readSTRV(dataSize)
        case "DESC": DESC = r.readSTRV(dataSize)
        case "LNAM": if LNAMs == nil { LNAMs = [LNAMField]() }; LNAMs.append(LNAMField(r, dataSize))
        default: return false
        }
        return true
    }
}
